src/aiida_codtools/calculations/functions/check_formula.py

```diff
@@ -0,0 +1,143 @@
+    """
+    Compare CIF formula against StructureData formula and report any inconsistency.
+    """
+    # pylint: disable=too-many-locals
+    from aiida.orm.nodes.data.cif import parse_formula as parse_formula_from_cif
+
+    report = ''
+    formula_s = structure.get_formula('hill', ' ')
+    formula_c = None
+    try:
+        assert len(cif.values.keys()) == 1, 'More than one CIF key.'
+        formula_c = get_formula_from_cif(cif)
+    except StarError:  # ignore unparsable CIF files (this should not happen)
+        report += ' | Unparsable CIF'
+        formula_c = cif.get_attribute('formulae')
+
+    if not formula_c:
+        # we cannot do any check without a formula... hope for the best
+        report += ' | No formula in CIF'
+        return report
+
+    report += f'cif [{formula_c}]  structure [{formula_s}]'
+
+    has_partial_occupancies = not structure.get_pymatgen().is_ordered
+    if has_partial_occupancies:
+        report += ' | Partial occupancies'
+
+    # get formula dictionaries {element: count}
+    formuladic_s = structure.get_pymatgen().composition.get_el_amt_dict()
+    formuladic_c = parse_formula_from_cif(formula_c)
+    # remove elements with zero occupancy
+    for key in [key for key, value in formuladic_c.items() if value == 0.]:
+        formuladic_c.pop(key)
+
+    # FIRST CHECK: find missing elements, ignore vacancies ('X')
+    # (symmetric difference of the sets -- contains elements not present in both sets)
+    missing_elements = (set(formuladic_s.keys()) ^ set(formuladic_c.keys())) - {'X'}
+    if missing_elements:
+        report += f' | Missing elements: {list(missing_elements)}'
+        raise MissingElementsError(report)
+        #return False, report
+
+    # SECOND CHECK: find inconsistent formulas
+    # ratios of all the elements should be the same
+    elements = formuladic_c.keys()
+    elcount_c = np.array([formuladic_c[key] for key in elements])
+    elcount_s = np.array([formuladic_s[key] for key in elements])
+
+    # ratios should be (all) > 1
+    # for integer occ., for each element the absolute tolerance on the ratio is 0.5 / (the smallest count)
+    # if the ratio difference is larger than this value, the structure can have more than one atom difference
+    if any(elcount_c > elcount_s):
+        ratios = elcount_c / elcount_s
+        atol = 0.499999 / np.ceil(elcount_s)
+    else:
+        ratios = elcount_s / elcount_c
+        atol = 0.499999 / np.ceil(elcount_c)
+
+    if has_partial_occupancies:
+        # for partial occ., compare ratios against the first one
+        # use an absolute tolerance of 0.05 + a 2 % relative tolerance
+        compare_with_ratio = ratios[0]
+        atol = 0.05
+        rtol = 0.02
+    else:
+        # for integer occ., compare ratios against the first one rounded to the nearest integer
+        # use only the absolute tolerance estimated above
+        compare_with_ratio = np.round(ratios[0])
+        rtol = 0.
```

Comment:
For _integer_ occupancies, I set the _relative_ tolerance to 0 and use only an absolute one.
The absolute tolerance is chosen to be the largest one possible: if the difference in ratios is larger than this tolerance, it means that there is at least a one-atom difference in the formulas.